### ISSUE TITLE
CA2000: Interlocked

### ISSUE THREAD

**Issue by skarllot** [dotnet/roslyn-analyzers#2681]
#### Analyzer package

[Microsoft.NetCore.Analyzers](https://www.nuget.org/packages/Microsoft.NetCore.Analyzers)

#### Package Version

v2.9.3

#### Diagnostic ID

[CA2000](https://docs.microsoft.com/en-us/visualstudio/code-quality/ca2000-dispose-objects-before-losing-scope?view=vs-2019)

#### Repro steps

Create the following code:

```csharp
class CustomDisposable : IDisposable
{
    Dispose() {}
}

class Test
{
    private CustomDisposable field1;

    private void NoWarning()
    {
        field1 = new CustomDisposable();
    }

    private void Warning1()
    {
        var temp = new CustomDisposable();
        Interlocked.Exchange(ref field1, temp)?.Dispose();
    }

    private void Warning2()
    {
        var temp = new CustomDisposable();
        Interlocked.CompareExchange(ref field1, temp, null);
    }
}
```

#### Expected behavior

Recognize thread safe assignments.

#### Actual behavior

Detect that temporary variable is being used and going out of scope without disposing.

**Comment by mavasani** [comment]
Yep, we need to special case all the Interlocked APIs and treat this as an assignment to the ref field in points to analysis.

**Comment by skarllot** [comment]
Should include `Volatile.Write` too
